### PR TITLE
Fix non-detected stale output file removal

### PR DESCRIPTION
**Pull request by wolfs**
We did not invalidate the file system mirror on stale output file deletion. This can cause up-to-date checks to snapshot the wrong state on disk.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CleanupStaleOutputsExecuter.java [hunk 1]
```diff
@@ -66,6 +68,7 @@ public void execute(final TaskInternal task, TaskStateInternal state, TaskExecut
             }
         }
         if (!filesToDelete.isEmpty()) {
+            taskOutputsGenerationListener.beforeTaskOutputsGenerated();
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CleanupStaleOutputsExecuter.java]: Should we rename the method to `beforeTaskOutputsChanged()`?

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy [hunk 2]
```diff
@@ -168,19 +168,23 @@ class StaleOutputIntegrationTest extends AbstractIntegrationSpec {
...
-        // recreate the output
+        changeGradleVersion()
         dirInBuildDir.createDir()
         staleFileInDir.touch()
         fileInBuildDir.touch()
         then:
         succeeds("myTask")
     }
 
+    // Now that we produce this, we can detect the situation where
```
Comment (lptr) [inline subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: What does "this" refer to here?

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy [hunk 3]
```diff
@@ -168,19 +168,23 @@ class StaleOutputIntegrationTest extends AbstractIntegrationSpec {
...
+    // someone builds with Gradle 4.3, then 4.2 and then 4.3 again.
+    // recreate the output
+    private void changeGradleVersion() {
```
Comment (lptr) [inline subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: Let's rename this method to something representing the goal to be achieved.

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: I changed the commen.

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: Renamed to `invalidateBuildOutputCleanupState`.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CleanupStaleOutputsExecuter.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/CleanupStaleOutputsExecuter.java]: Renamed to `beforeTaskOutputChanged()`.

**Comment by wolfs** [comment]
@lptr Please have another look.

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy [hunk 4]
```diff
@@ -436,18 +434,29 @@ class StaleOutputIntegrationTest extends AbstractIntegrationSpec {
...
-        changeGradleVersion()
-        run 'backup', 'restore'
+        /*
+        The whole setup is as follows:
+        - Both the restore and the backup task have an entry in the task history
+        - The output of the restore path is the input of the backup task.
+          This means when the backup task is up-to-date, then the contents of the output of the restore path is in the file system mirror.
+
+        If cleaning up stale output files does not invalidate the file system mirror, then the restore task would be up-to-date.
+         */
```
Comment (lptr) [inline subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: Style: I think we shouldn't be using `/* ... */`-style comments.

Comment (lptr) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: I'm still wondering if we could come up with an easier-to-understand test here if the two tasks were not locked in a circle. Like, we could trigger stale outputs just by creating a file among the outputs of the task between executions.

Comment (wolfs) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: We could create the file, but the file would not be in the file system mirror if no other task snapshots it. This is why I need the `backup` task to have the output of the `restore` task as in input.

Comment (wolfs) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputIntegrationTest.groovy]: I replaced `/*` by `//`.

**Comment by wolfs** [comment]
@lptr Please have another look.

**Pull request merged**